state-chain/pallets/cf-threshold-signature/src/lib.rs

```diff
@@ -1199,6 +1197,65 @@ pub mod pallet {
 	}
 }
 
+macro_rules! append_chain_to_name {
+	($name:ident) => {
+		match T::TargetChainCrypto::NAME {
+			"EVM" => concat!(stringify!($name), "EVM"),
+			"Polkadot" => concat!(stringify!($name), "Polkadot"),
+			"Bitcoin" => concat!(stringify!($name), "Bitcoin"),
+			"Solana" => concat!(stringify!($name), "Solana"),
+			_ => concat!(stringify!($name), "Other"),
+		}
+	};
+}
+
+impl<T, I> TypeInfo for pallet::CeremonyContext<T, I>
+where
+	T: Config<I>,
+	I: 'static,
+{
+	type Identity = Self;
+	fn type_info() -> Type {
+		Type::builder()
+			.path(Path::new(append_chain_to_name!(CeremonyContext), module_path!()))
+			.composite(
+				Fields::named()
+					.field(|f| {
+						f.ty::<RequestContext<T, I>>().name(append_chain_to_name!(RequestContext))
+					})
+					.field(|f| f.ty::<BTreeSet<T::ValidatorId>>().name("remaining_respondents"))
+					.field(|f| {
+						f.ty::<BTreeMap<T::ValidatorId, AuthorityCount>>().name("blame_counts")
+					})
+					.field(|f| f.ty::<BTreeSet<T::ValidatorId>>().name("candidates"))
+					.field(|f| f.ty::<EpochIndex>().name("epoch"))
+					.field(|f| {
+						f.ty::<<T::TargetChainCrypto as ChainCrypto>::AggKey>()
+							.name(append_chain_to_name!(Key))
+					})
+					.field(|f| f.ty::<ThresholdCeremonyType>().name("threshold_ceremony_type")),
+			)
+	}
+}
+
+impl<T, I> TypeInfo for pallet::RequestContext<T, I>
+where
+	T: Config<I>,
+	I: 'static,
+{
+	type Identity = Self;
+	fn type_info() -> Type {
+		Type::builder()
+			.path(Path::new(append_chain_to_name!(RequestContext), module_path!()))
+			.composite(
+				Fields::named()
+					.field(|f| f.ty::<RequestId>().name("request_id"))
+					.field(|f| f.ty::<AttemptCount>().name("attempt_count"))
+					.field(|f| f.ty::<PayloadFor<T, I>>().name(append_chain_to_name!(PayloadFor))),
```

Comment:
Sorry, another comment... 

I don't understand why we need to use this `append_chain_to_name` on the field names... surely it's only needed on the path? 

If we want to resolve the type `PayloadFor<T, I>` then we need to apply the same treatment to the `PayloadFor<T, I>` *types* and somehow direct the type_info to resolve that type.

Reply:
The only way to direct type_info on how to resolve a type is via its name. If we don't append the chain name to PayloadFor, then there will be multiple types called "PayloadFor" and the recursive type resolution will fail again.

Reply:
my previous message was not correct. I added separate "field names" and "type names" now